crypto/hmac/hmac_s390x.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "crypto/s390x_arch.h"
+#include "hmac_local.h"
+#include "openssl/obj_mac.h"
+#include "openssl/evp.h"
+
+#ifdef OPENSSL_HMAC_S390X
+
+static int s390x_fc_from_md(const EVP_MD *md)
+{
+    int fc;
+
+    switch (EVP_MD_get_type(md)) {
+    case NID_sha224:
+        fc = S390X_HMAC_SHA_224;
+        break;
+    case NID_sha256:
+        fc = S390X_HMAC_SHA_256;
+        break;
+    case NID_sha384:
+        fc = S390X_HMAC_SHA_384;
+        break;
+    case NID_sha512:
+        fc = S390X_HMAC_SHA_512;
+        break;
+    default:
+        return 0;
+    }
+
+    if ((OPENSSL_s390xcap_P.kmac[1] & S390X_CAPBIT(fc)) == 0)
+        return 0;
+
+    return fc;
+}
+
+static void s390x_call_kmac(HMAC_CTX *ctx, const unsigned char *in, size_t len)
+{
+    unsigned int fc = ctx->plat.s390x.fc;
+
+    if (ctx->plat.s390x.ikp)
+        fc |= S390X_KMAC_IKP;
+
+    if (ctx->plat.s390x.iimp)
+        fc |= S390X_KMAC_IIMP;
+
+    switch (ctx->plat.s390x.fc) {
+    case S390X_HMAC_SHA_224:
+    case S390X_HMAC_SHA_256:
+        ctx->plat.s390x.param.hmac_224_256.imbl += ((uint64_t)len * 8);
+        break;
+    case S390X_HMAC_SHA_384:
+    case S390X_HMAC_SHA_512:
+        ctx->plat.s390x.param.hmac_384_512.imbl += ((uint128_t)len * 8);
+        break;
+    default:
+        break;
+    }
+
+    s390x_kmac(in, len, fc, &ctx->plat.s390x.param);
+
+    ctx->plat.s390x.ikp = 1;
+}
+
+int s390x_HMAC_init(HMAC_CTX *ctx, const void *key, int key_len, ENGINE *impl)
```

Comment:
I have a question: is it necessary to pass `impl` argument? If I understand things right the `ENGINE` API is deprecated. The function `EVP_DigestInitEx()` which consumes the `impl` argument ignores it.

So getting rid of `impl` argument in `s390x_HMAC_init()` and passing `NULL` to `EVP_DigestInitEx()` would make code to look bit better with respect to documentation.

Reply:
I doubt that `EVP_DigestInitEx()` really ignores the `impl` argument. 

Although engines are deprecated, you can still configure an engine even with OpenSSL 3.x. And if that engine implements the digest algorithm used here, then it would be used on these calls, as far as I can tell.

Reply:
> I doubt that `EVP_DigestInitEx()` really ignores the `impl` argument.
> 
> Although engines are deprecated, you can still configure an engine even with OpenSSL 3.x. And if that engine implements the digest algorithm used here, then it would be used on these calls, as far as I can tell.

Sorry, you are absolutely right. I got confused by accidentaly poking to libressl manpage instead of OpenSSL.